### PR TITLE
llvm 3.9: Download and build Polly

### PR DESCRIPTION
**Pull request by spacekitteh**
###### Motivation for this change
#20890

###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by spacekitteh** [comment]
Well, this is going to take a while to check and build...

**Comment by spacekitteh** [comment]
Well, it looks like the travis instances don't have enough RAM to build it.

**Comment by spacekitteh** [comment]
So, I get this error:

```bashCMakeFiles/clang.dir/cc1_main.cpp.o: In function `cc1_main(llvm::ArrayRef<char const*>, char const*, void*)': cc1_main.cpp:(.text._Z8cc1_mainN4llvm8ArrayRefIPKcEES2_Pv+0x35b): undefined reference to `polly::initializePollyPasses(llvm::PassRegistry&)'```

**Comment by spacekitteh** [comment]
If I can't figure out how to appropriately patch the CMakeLists.txt, I don't think I can get around building polly twice: once in llvm.nix, and again in clang.nix.

I think this might be an example of the perfect use-case of multiple build outputs. Polly has to be in the llvm/tools/polly dir upon building clang in order to statically link it, unless the cmake file is patched.

File: pkgs/development/compilers/llvm/3.9/clang/default.nix [hunk 1]
```diff
@@ -26,7 +28,7 @@ let
 
     postPatch = ''
       sed -i -e 's/Args.hasArg(options::OPT_nostdlibinc)/true/' lib/Driver/Tools.cpp
-      sed -i -e 's/DriverArgs.hasArg(options::OPT_nostdlibinc)/true/' lib/Driver/ToolChains.cpp
+      sed -i -e 's/DriverArgs.hasArg(options::OPT_nostdlibinc)/true/' lib/Driver/ToolChains.cpp
```
Comment (copumpkin) [inline pkgs/development/compilers/llvm/3.9/clang/default.nix]: whitespace!

**Comment by copumpkin** [comment]
@shlevy has also worked a fair amount on LLVM, if I remember correctly

**Comment by spacekitteh** [comment]
So, I currently get the following error when using clang: `: CommandLine Error: Option 'aarch64-ccmp' registered more than once! LLVM ERROR: inconsistency in registered CommandLine options`

Possibly related: https://llvm.org/bugs/show_bug.cgi?id=24953

It might be because it's linking against a .so instead of a .a?

**Comment by spacekitteh** [comment]
It looks like LLVM_BUILD_LLVM_DYLIB is broken, actually.

File: pkgs/development/compilers/llvm/3.9/clang/default.nix [hunk 2]
```diff
@@ -17,7 +17,10 @@ let
 
     cmakeFlags = [
       "-DCMAKE_CXX_FLAGS=-std=c++11"
+      "-DWITH_POLLY:Bool=true"
+      "-DLINK_POLLY_INTO_TOOLS:Bool=true"            "-DLLVM_LINK_LLVM_DYLIB:Bool=true"
     ] ++
+#    (stdenv.lib.optional llvm.enableSharedLibraries "-DLLVM_LINK_LLVM_DYLIB:Bool=true") ++
```
Comment (spacekitteh) [inline pkgs/development/compilers/llvm/3.9/clang/default.nix]: how can i do this? That is, how to reference the input argument to a build dependency?

Comment (spacekitteh) [inline pkgs/development/compilers/llvm/3.9/clang/default.nix]: eh, doesn't matter anyway. it doesn't fix the problem.

**Comment by spacekitteh** [comment]
There's this message, too:
```bash
WARNING: substitution variables should be valid bash names, "ccLDFlags+" isn't and therefore was skipped; it might be caused by multi-line phases in variables - see #14907 for details.
```

**Comment by spacekitteh** [comment]
See [here](https://github.com/NixOS/nixpkgs/blob/7077a270bf625b320ab61e3951aa19c9d514b5dc/pkgs/build-support/cc-wrapper/default.nix#L125).

I think there might be a problem with the spacing. I'll try changing that.

**Comment by spacekitteh** [comment]
I think it's a problem in [this bit of code](https://github.com/NixOS/nixpkgs/blob/master/pkgs/stdenv/generic/setup.sh#L419).

It thinks that `ccCFlags+=` is a variable being assigned called `ccCFlags+`, as opposed to the variable`ccCFlags` being updated.

**Comment by spacekitteh** [comment]
Damn, #ddb1321 didn't fix it. Anyone have any ideas?

**Comment by dtzWill** [comment]
By the way, did the problems motivating https://github.com/NixOS/nixpkgs/pull/20941/commits/31bdc2f770b28e7bfc46d24f665dad3c7c888810 ever get reported/fixed in Nixpkgs master?

Also--what's the status of this? Are there outstanding problems with this
(other than probably needing a squash/rebase to cleanup the history)?

**Comment by vcunat** [comment]
I don't think it's been fixed, and I've never seen that problem causing anything more than warnings.

**Comment by Mic92** [comment]
This needs to be rewritten for newer llvm expressions

**Pull request closed without merging**